homeassistant/components/wemo/__init__.py

```diff
@@ -187,15 +187,43 @@ def __init__(self, hass: HomeAssistant, wemo_dispatcher: WemoDispatcher) -> None
         self._wemo_dispatcher = wemo_dispatcher
         self._stop = None
         self._scan_delay = 0
+        self._upnp_entries = set()
+
+    async def async_add_from_upnp_entry(self, entry: pywemo.ssdp.UPNPEntry) -> None:
+        """Create a WeMoDevice from an UPNPEntry and add it to the dispatcher.
+
+        Uses the self._upnp_entries set to avoid interrogating the same device
+        multiple times.
+        """
+        if entry in self._upnp_entries:
+            return
+        try:
+            device = await self._hass.async_add_executor_job(
```

Comment:
How many devices can they have at a time?

Reply:
I guess it's limited by the size of the subnet? So a max of 252 for a typical class C network that also includes Home Assistant and a default gateway.

I've been testing with 22 devices.

Reply:
If they have more than 38 devices it will tie up every executor thread and everything that needs the executor will have to wait.

Can we make this async or limit to 3 at a time?￼

Reply:
Thanks for that. I hadn't realized the limit on the executors. This makes sense, and it looks like there is a `gather_with_concurrency` helper that can be used here. I'll give that a try and test things out.

Reply:
All set and tested. I also applied the same change to the initialization of the static config entries.